File: typescript/web_core/src/processing/adapters/factory.ts
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {A2uiValidationError} from '../../errors.js';
import {normalizeVersionString, toCanonicalVersion} from '../../common/semver.js';
import {
  ProtocolVersion,
  VersionAdapter,
  VersionAdapterResolver,
  registerKnownActionsProvider,
} from './base.js';
import {V0Point8Adapter} from './v0_8.js';
import {V0Point9Adapter} from './v0_9.js';
import {V1Point0Adapter} from './v1_0.js';

/**
 * Resolves version adapters for protocol specification versions.
 */
export class VersionAdapterFactory implements VersionAdapterResolver {
  private readonly adapters = new Map<string, VersionAdapter>([
    ['0.8', new V0Point8Adapter()],
    ['0.9', new V0Point9Adapter()],
    ['0.9.1', new V0Point9Adapter()],
    ['1.0', new V1Point0Adapter()],
  ]);

  /**
   * Returns the aggregated set of all action keys supported by all registered adapters.
   *
   * @return A set of all supported action key strings.
   */
  getAllKnownActions(): ReadonlySet<string> {
    const actions = new Set<string>();
    for (const adapter of this.adapters.values()) {
      if (adapter.validActions) {
        for (const action of adapter.validActions) {
          actions.add(action);
        }
      }
    }
    return actions;
  }

  /**
   * Dynamically registers a version adapter on this factory instance.
   *
   * @param adapter The version adapter instance to register.
   */
  registerAdapter(adapter: VersionAdapter): void {
    const key = toCanonicalVersion(adapter.version) ?? normalizeVersionString(adapter.version);
    this.adapters.set(key, adapter);
  }

  /**
   * Resolves the version adapter for the specified version string from this factory instance.
   *
   * @param version The protocol version string (e.g. 'v1.0').
   * @return The matching version adapter.
   * @throws A2uiValidationError if the version string is unsupported.
   */
  getAdapter(version: ProtocolVersion | string): VersionAdapter {
    const key = toCanonicalVersion(version) ?? normalizeVersionString(version);
    const adapter = this.adapters.get(key);
    if (!adapter) {
      const supported = Array.from(
        new Set(Array.from(this.adapters.keys()).map(k => (k.startsWith('v') ? k : `v${k}`))),
      )
        .sort()
        .join(', ');
      throw new A2uiValidationError(
        `[VersionAdapterFactory] Unsupported protocol version '${version}'. Supported versions: ${supported}.`,
      );
    }
    return adapter;
  }

  /**
   * Resolves the version adapter directly from an incoming message payload using this factory instance.
   *
   * @param payload The raw JSON message payload.
   * @returns The resolved version adapter.
   * @throws A2uiValidationError if the payload is missing a valid 'version' string.
   */
  resolveFromPayload(payload: unknown): VersionAdapter {
    const item = Array.isArray(payload) ? payload[0] : payload;
    if (typeof item === 'object' && item !== null) {
      if ('messages' in item && Array.isArray((item as any).messages)) {
        return this.resolveFromPayload((item as any).messages);
      }
      if ('version' in item) {
        const ver = (item as {version: unknown}).version;
        if (typeof ver !== 'string') {
          throw new A2uiValidationError(
            `[VersionAdapterFactory] Message payload is missing a valid 'version' string: 'version' property must be a string, got ${typeof ver}.`,
          );
        }
        return this.getAdapter(ver);
      }
      if (
        'beginRendering' in item ||
        'surfaceUpdate' in item ||
        'dataModelUpdate' in item ||
        'deleteSurface' in item
      ) {
        return this.getAdapter('v0.8');
      }
    }
    throw new A2uiValidationError(
      "[VersionAdapterFactory] Message payload is missing a valid 'version' string.",
    );
  }

  /**
   * Registers a version adapter on the default singleton factory instance.
   *
   * @param adapter Version adapter instance to register.
   */
  static registerAdapter(adapter: VersionAdapter): void {
    defaultVersionAdapterFactory.registerAdapter(adapter);
  }

  /**
   * Resolves a version adapter for the specified version from the default singleton factory instance.
   *
   * @param version Protocol version string.
   * @return Matching version adapter.
   */
  static getAdapter(version: ProtocolVersion | string): VersionAdapter {
    return defaultVersionAdapterFactory.getAdapter(version);
  }

  /**
   * Resolves a version adapter directly from a payload using the default singleton factory instance.
   *
   * @param payload Raw JSON message payload.
   * @returns Resolved version adapter.
   */
  static resolveFromPayload(payload: unknown): VersionAdapter {
    return defaultVersionAdapterFactory.resolveFromPayload(payload);
  }

  /**
   * Returns the aggregated set of all action keys supported by all registered adapters
   * from the default singleton factory instance.
   *
   * @return A set of all supported action key strings.
   */
  static getAllKnownActions(): ReadonlySet<string> {
    return defaultVersionAdapterFactory.getAllKnownActions();
  }
}

/** Default singleton version adapter factory instance. */
export const defaultVersionAdapterFactory = new VersionAdapterFactory();
registerKnownActionsProvider(() => defaultVersionAdapterFactory.getAllKnownActions());

File: python/a2ui_core/src/a2ui/core/processing/adapters/factory.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from collections.abc import Mapping, Sequence
from typing import Any
from .base import VersionAdapter
from .v0_8 import V0Point8Adapter
from .v0_9 import V0Point9Adapter
from .v1_0 import V1Point0Adapter
from ...exceptions import A2uiErrorDetail, A2uiValidationError
from ...schema import AgentToRendererMessage, ProtocolVersion

DEFAULT_PROTOCOL_VERSION: ProtocolVersion = ProtocolVersion.V0_9


class VersionAdapterFactory:
    """Resolves version adapters for protocol specification versions."""

    _adapters: dict[ProtocolVersion | str, VersionAdapter] = {
        ProtocolVersion.V0_8: V0Point8Adapter(),
        ProtocolVersion.V0_9: V0Point9Adapter(),
        ProtocolVersion.V0_9_1: V0Point9Adapter(),
        ProtocolVersion.V1_0: V1Point0Adapter(),
    }

    @classmethod
    def register_adapter(cls, adapter: VersionAdapter) -> None:
        """Dynamically registers a version adapter."""
        cls._adapters[adapter.version] = adapter

    @classmethod
    def get_adapter(cls, version: ProtocolVersion | str) -> VersionAdapter:
        """Resolves the version adapter for the specified protocol version enum or string."""
        adapter = None
        if isinstance(version, str):
            parsed_ver = cls._parse_version(version)
            if parsed_ver:
                adapter = cls._adapters.get(parsed_ver)
            if not adapter:
                adapter = cls._adapters.get(version)
        elif isinstance(version, ProtocolVersion):
            adapter = cls._adapters.get(version)

        if not adapter:
            supported = ", ".join(
                v.value for v in cls._adapters.keys() if hasattr(v, "value")
            )
            raise A2uiValidationError(
                f"[VersionAdapterFactory] Unsupported protocol version '{version}'."
                f" Supported versions: {supported}."
            )
        return adapter

    @classmethod
    def resolve_from_payload(
        cls,
        payload: (
            AgentToRendererMessage
            | Sequence[AgentToRendererMessage]
            | Mapping[str, Any]
            | Sequence[Mapping[str, Any]]
        ),
    ) -> VersionAdapter:
        """Resolves the version adapter directly from an incoming message payload."""
        if payload is None or isinstance(payload, (int, float, str)):
            raise A2uiValidationError(
                "[VersionAdapterFactory] Message payload is missing a valid 'version'"
                " string.",
                details=[
                    A2uiErrorDetail(
                        path="messages.0.version",
                        code="missing_field",
                        message="Missing required version field",
                    )
                ],
            )

        raw_payload: Any = payload
        if hasattr(raw_payload, "model_dump"):
            raw_payload = raw_payload.model_dump(by_alias=True, exclude_none=True)

        if isinstance(raw_payload, list):
            if not raw_payload:
                raise A2uiValidationError(
                    "[VersionAdapterFactory] Message payload is missing a valid"
                    " 'version' string.",
                    details=[
                        A2uiErrorDetail(
                            path="messages.0.version",
                            code="missing_field",
                            message="Missing required version field",
                        )
                    ],
                )
            for idx, item in enumerate(raw_payload):
                raw_item: Any = item
                if hasattr(raw_item, "model_dump"):
                    raw_item = raw_item.model_dump(by_alias=True, exclude_none=True)
                if isinstance(raw_item, dict):
                    v = raw_item.get("version")
                    if not v or not isinstance(v, str):
                        if not any(
                            k in raw_item
                            for k in (
                                "beginRendering",
                                "surfaceUpdate",
                                "dataModelUpdate",
                                "deleteSurface",
                            )
                        ):
                            raise A2uiValidationError(
                                "[VersionAdapterFactory] Message payload is missing a"
                                " valid 'version' string.",
                                details=[
                                    A2uiErrorDetail(
                                        path=f"messages.{idx}.version",
                                        code="missing_field",
                                        message="Missing required version field",
                                    )
                                ],
                            )
                    return cls._resolve_from_single_action(raw_item)

        if isinstance(raw_payload, dict):
            if "messages" in raw_payload and isinstance(raw_payload["messages"], list):
                return cls.resolve_from_payload(raw_payload["messages"])
            v = raw_payload.get("version")
            if not v or not isinstance(v, str):
                if not any(
                    k in raw_payload
                    for k in (
                        "beginRendering",
                        "surfaceUpdate",
                        "dataModelUpdate",
                        "deleteSurface",
                    )
                ):
                    raise A2uiValidationError(
                        "[VersionAdapterFactory] Message payload is missing a valid"
                        " 'version' string.",
                        details=[
                            A2uiErrorDetail(
                                path="messages.0.version",
                                code="missing_field",
                                message="Missing required version field",
                            )
                        ],
                    )
            return cls._resolve_from_single_action(raw_payload)

        raise A2uiValidationError(
            "[VersionAdapterFactory] Message payload is missing a valid 'version'"
            " string.",
            details=[
                A2uiErrorDetail(
                    path="messages.0.version",
                    code="missing_field",
                    message="Missing required version field",
                )
            ],
        )

    @classmethod
    def _resolve_from_single_action(cls, item: dict[str, Any]) -> VersionAdapter:
        """Resolves version adapter from a single message dictionary if explicit version or action keys match."""
        if "version" in item and isinstance(item["version"], str):
            ver_str = item["version"]
            ver_enum = cls._parse_version(ver_str)
            if not ver_enum:
                supported = ", ".join(
                    v.value for v in cls._adapters.keys() if hasattr(v, "value")
                )
                raise A2uiValidationError(
                    f"[VersionAdapterFactory] Unsupported protocol version '{ver_str}'."
                    f" Supported versions: {supported}."
                )
            return cls.get_adapter(ver_enum)
        if any(
            k in item
            for k in (
                "beginRendering",
                "surfaceUpdate",
                "dataModelUpdate",
                "deleteSurface",
            )
        ):
            return cls.get_adapter(ProtocolVersion.V0_8)
        raise A2uiValidationError(
            "[VersionAdapterFactory] Message payload is missing a valid 'version'"
            " string.",
            details=[
                A2uiErrorDetail(
                    path="messages.0.version",
                    code="missing_field",
                    message="Missing required version field",
                )
            ],
        )

    @classmethod
    def _parse_version(cls, version_str: str) -> ProtocolVersion | None:
        """Parses a version string into a ProtocolVersion enum.

        Args:
            version_str: The version string to parse.

        Returns:
            The matched ProtocolVersion enum member, or None if unrecognized.
        """
        from ...common.semver import to_canonical_version

        canonical = to_canonical_version(version_str)
        if canonical:
            canonical_map = {
                "0.8": ProtocolVersion.V0_8,
                "0.9": ProtocolVersion.V0_9,
                "0.9.1": ProtocolVersion.V0_9_1,
                "1.0": ProtocolVersion.V1_0,
            }
            if canonical in canonical_map:
                return canonical_map[canonical]

        clean = (
            f"v{version_str[1:]}"
            if version_str.startswith(("v", "V"))
            else f"v{version_str}"
        )
        try:
            return ProtocolVersion(clean)
        except ValueError:
            return None
